### PR TITLE
Remote Server Login

### PR DESCRIPTION
**Pull request by kress95**
dancing because remote server login

> dancing because remote server login

Tasks:
---------

- [x] Remote Server Access
  - ~~Make Login/Crack/Scan global~~ (won't do it that way)
  - [x] Make Hacking Toolkit widget 
  - [x] Integrate Hacking Toolkit to Some Pages
  - [x] Bruteforce hook
  - [x] Clicking on login button dispaches to Web...
  - [x] ...that triggers a Dispatch Join and...
  - [x] ...a Handler to receive joined server event.
  - [x] Request bootstrap (actually done but not hooked, lets discuss #239)
  - [x] Insert bootstrap data (actually done but not hooked, lets discuss #239)
  - [x] Dispatch Success/Error back to browser page
  - ~~Change the page on success, visually report the error on failure~~ (delegated to @pedrohlc)
  - [x] Change the endpoint to the invaded server

Effects:
---------

- [x] Reimplement Web model (on Game)
- [x] Move requester type to the model (and update everyone who's using it)
- [x] Re add events to Web
- [x] Include topic and json on Websocket `joined`/`joinFailed` reports
- [x] #240 Refactor Websocket Driver
- [x] Rename Bootstrap fields.
- [x] Include gatewayId on server join.
- [x] Include nip on DNS request.
- [x] Rewrite (yet another time DNS) request.
- [x] Adapt page models to the new metadata format.
- [x] Update tests.
- [x] Web service files were confusing, tried fixing it and had trouble with cyclical dependencies (wasting a considerable amount of time).
- [x] Better API for creating crackers with unspecified files
- [x] Split Widget in two files (to avoid depending on Html.CssHelpers) *
- [x] Cut tab report code on another commit (we'll use model checking instead)

`*` Not sure if this is the proper solution.

Discuss:
---------

- [x] Discuss about what happens when you login on a server and changes the endpoint before receiving responses from the server.
- [x] Discuss about replacing bootstrap with a generic onJoin... See [this link](https://github.com/HackerExperience/HEBorn/issues/239) [**accepted**|~~declined~~]

Next PRs:
-----------

1. Rewrite bootstrap to use the proposed #239 format.
2. Remote control, include opening apps on endpoint context, file upload and these kind of things.

---

🔝 PR DESCRIPTION

---
This change is [Reviewable](https://reviewable.io/reviews/hackerexperience/heborn/241)

### CODE REVIEW

**Comment by DUFC-Liam** [comment]
Imo this is a good feature and can get simple implementation on soon 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/heborn/241#-:-KugP_wBvE-nEbEaYthO:b3ouzc7)*

**Comment by PedroHLC** [comment]
> (@PedroHLC gibe support plox)
👍

File: src/Game/Web/DNS.elm [hunk 1]
```diff
@@ -1,18 +1,10 @@
 module Game.Web.DNS exposing (..)
 
-import Game.Meta.Types exposing (Context(..))
 import Game.Web.Types exposing (..)
+import Game.Web.Models exposing (..)
 
 
 type Response
```
Comment (PedroHLC) [inline src/Game/Web/DNS.elm]: This can go to Models too

File: src/Game/Web/Models.elm [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import Game.Meta.Types exposing (Context(..))
+
+
+type alias Model =
+    { loadingPages : LoadingPages
+    }
+
+
+type alias LoadingPages =
+    Dict Servers.ID Requester
```
Comment (PedroHLC) [inline src/Game/Web/Models.elm]: This is insane, explain it to me...

File: src/Apps/Browser/Pages/Models.elm [hunk 3]
```diff
@@ -35,7 +35,7 @@ type Model
 
 
 initialModel : Web.Site -> Model
-initialModel { url, type_ } =
+initialModel ({ url, type_ } as site) =
```
Comment (PedroHLC) [inline src/Apps/Browser/Pages/Models.elm]: site includes unnecessary type, page doesn't need to receive this intel, better to it receive only ```url content meta``

File: src/Game/Web/Requests/DNS.elm [hunk 4]
```diff
@@ -74,110 +78,150 @@ errorMessage url str =
...
+encoder url networkId =
     Encode.object
-        [ ( "url", Encode.string url )
+        [ ( "network_id", Encode.string networkId )
+        , ( "address", Encode.string url )
         ]
 
 
 decoder : Url -> Decoder Response
 decoder url =
```
Comment (PedroHLC) [inline src/Game/Web/Requests/DNS.elm]: This change became completely unnecessary, see [Helix#270](https://github.com/HackerExperience/Helix/issues/270l)

**Comment by PedroHLC** [comment]
I'm overly satisfied with this PR, feels like everything was greatly discussed during production!

---

Reviewed 5 of 17 files at r1, 9 of 19 files at r2, 15 of 23 files at r3, 13 of 13 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/Game/Update.elm, line 224 at r4](https://reviewable.io:443/reviews/hackerexperience/heborn/241#-Kv49M7muJZpbHOmgFij:-Kv49M7muJZpbHOmgFik:b-ib47kr) ([raw file](https://github.com/hackerexperience/heborn/blob/fc2961ab60ad2561f3d679e9c7ad34b0eba16fda/src/Game/Update.elm#L224)):*
> ```Elm
> 
>         joinServer gatewayId list =
>             -- TODO: include endpoint join
> ```

Found this TODO

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/heborn/241#-:-Kv4A2rDqNh4PKa1yRfS:br0prkj)*

**Comment by PedroHLC** [comment]
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/hackerexperience/heborn/241)*

**Pull request merged**